memory/memory/client.py

```diff
@@ -0,0 +1,79 @@
+from asyncinit import asyncinit
+import aiohttp
+import concurrent
+
+from hailtop.auth import service_auth_headers
+from hailtop.config import get_deploy_config
+from hailtop.google_storage import GCS
+from hailtop.tls import get_context_specific_ssl_client_session
+from hailtop.utils import async_to_blocking, request_retry_transient_errors
+
+@asyncinit
+class AsyncMemoryClient:
+    async def __init__(self, billing_project, fs=None, deploy_config=None, session=None,
+                       headers=None, _token=None):
+        self.billing_project = billing_project
+
+        if not deploy_config:
+            deploy_config = get_deploy_config()
+
+        self.url = deploy_config.base_url('memory')
+
+        if session is None:
+            session = get_context_specific_ssl_client_session(
+                raise_for_status=True,
+                timeout=aiohttp.ClientTimeout(total=60))
+        self._session = session
+
+        if fs is None:
+            fs = GCS(blocking_pool=concurrent.futures.ThreadPoolExecutor(), project=billing_project)
+        self._fs = fs
+
+        h = {}
+        if headers:
+            h.update(headers)
+        if _token:
+            h['Authorization'] = f'Bearer {_token}'
+        else:
+            h.update(service_auth_headers(deploy_config, 'memory'))
+        self._headers = h
+
+    async def _get_file_if_exists(self, filename):
+        request = {
+            'billing_project': self.billing_project,
+            'filename': filename}
+        try:
+            response = await request_retry_transient_errors(
+                self._session, 'POST',
+                self.url + '/getfile', json=request, headers=self._headers)
+            async with response:
+                if response.status == 200:
+                    return await response.read()
+        except aiohttp.ClientResponseError as e:
+            if e.status == 404:
+                return None
+            raise e
+
+    async def read_file(self, filename):
+        data = await self._get_file_if_exists(filename)
```

Comment:
I think this should hit Google and verify the object exists before checking the cache.  It should pass the etag to the cache, and that (+/- the object key/path) should be the lookup key.

Reply:
Can we benchmark getting the etags/metadata?

Reply:
So the way I ended up implementing this:
- client does an etag check for existence/version info, passes etag info to server
- server checks to see if file exists in cache and if the etag matches; if so, return file
- if file doesn't exist or etag does not match, add file to read queue
- when the server reads the file, it also gets the etag and stores both into the cache (to deal with the case where the file could be updated between when the client sends the request and the server reads the file)

Right now I'm getting the etag and the file data in separate calls since that was the easiest way to implement it on the current google storage interface; I'll unify the calls when we lift this to the new API, as discussed.